src/python/utils/diff_statistics.py

```diff
@@ -0,0 +1,167 @@
+        right_sha (str): The right sha.
+        cache_dir (Path): The path to the cache directory.
+    Returns:
+        Set[str]: A set containing the diff result.
+    """
+    command = f"git merge-base {left_sha} {right_sha}"
+    base_sha = repo.run_command(command)[0].strip()
+    left_right_files = get_diff_files_branches(repo, left_sha, right_sha)
+    base_right_files = get_diff_files_branches(repo, base_sha, right_sha)
+    base_left_files = get_diff_files_branches(repo, base_sha, left_sha)
+
+    # Check that at least one java file is contained in all 3 diffs
+    common_files = left_right_files & base_right_files & base_left_files
+
+    return common_files
+
+
+def diff_contains_java_file(
+    repo: Repository, left_sha: str, right_sha: str
+) -> Union[bool, None]:
+    """
+    Computes the diff between two branches using git diff.
+    Args:
+        repo (Repository): The repository object.
+        left_sha (str): The left sha.
+        right_sha (str): The right sha.
+    Returns:
+        bool: True if the diff contains a java file, False otherwise.
+    """
+    try:
+        merge_diff = get_diff_files_merge(repo, left_sha, right_sha)
+    except Exception as e:
+        logger.error(
+            f"diff_contains_java_file: {left_sha} {right_sha} {repo.repo_slug} {e}"
+        )
+        return None
+    return any(file.endswith(".java") for file in merge_diff)
+
+
+def diff_contains_non_java_file(
+    repo: Repository, left_sha: str, right_sha: str
+) -> Union[bool, None]:
+    """
+    Computes the diff between two branches using git diff.
+    Args:
+        repo (Repository): The repository object.
+        left_sha (str): The left sha.
+        right_sha (str): The right sha.
+    Returns:
+        bool: True if the diff contains a non-Java file, False otherwise.
+    """
+    try:
+        merge_diff = get_diff_files_branches(repo, left_sha, right_sha)
+    except Exception as e:
+        logger.error(
+            f"diff_contains_non_java_file: {left_sha} {right_sha} {repo.repo_slug} {e}"
+        )
+        return None
+    return any(not file.endswith(".java") for file in merge_diff)
+
+
+def compute_num_diff_files(
+    repo: Repository, left_sha: str, right_sha: str
+) -> Union[int, None]:
+    try:
+        diff_file = get_diff_files_branches(repo, left_sha, right_sha)
+    except Exception as e:
+        logger.error(
+            f"compute_num_diff_files: {left_sha} {right_sha} {repo.repo_slug} {e}"
+        )
+        return None
+    return len(diff_file)
+
+
+def compute_num_diff_lines(
+    repo: Repository, left_sha: str, right_sha: str
+) -> Union[int, None]:
+    try:
+        diff = get_diff(repo, left_sha, right_sha, None)
+    except Exception as e:
+        logger.error(
+            f"compute_num_diff_lines: {left_sha} {right_sha} {repo.repo_slug} {e}"
+        )
+        return None
+    return sum(1 for line in diff.splitlines() if line.startswith(("+ ", "- ")))
```

Comment:
The regular expressions are wrong.  They should be "+[^+]" and "-[^-]" but even that is wrong for two reasons.
 * A line might start with + or -, in which case the diff line would start with "++" or "--" (we'll ignore this detail).
 * A line might be empty, in which case the diff line would be "+" or "-" in toto.  This is important to fix.

So, something like "+($|[^+])" might work, depending on the behavior of $ in Python and whether the initial + needs to be quoted.  Please test.

Reply:
Oh, wait, this is "startswith".  Does that handle regular expressions?  if not, then:
 * startswith + or -
 * does not startwith ++ or --

is enough.